d2/src/main/java/com/linkedin/d2/xds/balancer/XdsTogglingLoadBalancerFactory.java

```diff
@@ -0,0 +1,180 @@
+    _d2ClientJmxManager = d2ClientJmxManager;
+    _deterministicSubsettingMetadataProvider = deterministicSubsettingMetadataProvider;
+    _failoutConfigProviderFactory = failoutConfigProviderFactory;
+    _canaryDistributionProvider = canaryDistributionProvider;
+  }
+
+  public TogglingLoadBalancer create(ScheduledExecutorService executorService, XdsToD2PropertiesAdaptor xdsAdaptor)
+  {
+    PropertyEventBus<ClusterProperties> clusterBus = new PropertyEventBusImpl<>(executorService);
+    PropertyEventBus<ServiceProperties> serviceBus = new PropertyEventBusImpl<>(executorService);
+    PropertyEventBus<UriProperties> uriBus = new PropertyEventBusImpl<>(executorService);
+
+    FileStore<ClusterProperties> fsClusterStore =
+        createFileStore(FileSystemDirectory.getClusterDirectory(_fsd2DirPath), new ClusterPropertiesJsonSerializer());
+    _d2ClientJmxManager.setFsClusterStore(fsClusterStore);
+
+    FileStore<ServiceProperties> fsServiceStore =
+        createFileStore(FileSystemDirectory.getServiceDirectory(_fsd2DirPath, _d2ServicePath),
+            new ServicePropertiesJsonSerializer());
+    _d2ClientJmxManager.setFsServiceStore(fsServiceStore);
+
+    FileStore<UriProperties> fsUriStore =
+        createFileStore(_fsd2DirPath + File.separator + "uris", new UriPropertiesJsonSerializer());
+    _d2ClientJmxManager.setFsUriStore(fsUriStore);
+
+    // This ensures the filesystem store receives the events from the event bus so that
+    // it can keep a local backup.
+    clusterBus.register(fsClusterStore);
+    serviceBus.register(fsServiceStore);
+    uriBus.register(fsUriStore);
+
+    XdsToClusterPropertiesPublisher clusterPropertiesPublisher = new XdsToClusterPropertiesPublisher(xdsAdaptor);
+    XdsToServicePropertiesPublisher servicePropertiesPublisher = new XdsToServicePropertiesPublisher(xdsAdaptor);
+    XdsToUriPropertiesPublisher uriPropertiesPublisher = new XdsToUriPropertiesPublisher(xdsAdaptor);
+
+    TogglingPublisher<ClusterProperties> clusterToggle =
+        new TogglingPublisher<>(clusterPropertiesPublisher, fsClusterStore, clusterBus);
+    TogglingPublisher<ServiceProperties> serviceToggle =
+        new TogglingPublisher<>(servicePropertiesPublisher, fsServiceStore, serviceBus);
+    TogglingPublisher<UriProperties> uriToggle = new TogglingPublisher<>(uriPropertiesPublisher, fsUriStore, uriBus);
+
+    SimpleLoadBalancerState state =
+        new SimpleLoadBalancerState(executorService, uriBus, clusterBus, serviceBus, _clientFactories,
+            _loadBalancerStrategyFactories, _sslContext, _sslParameters, _isSSLEnabled, _partitionAccessorRegistry,
+            _sslSessionValidatorFactory, _deterministicSubsettingMetadataProvider, _canaryDistributionProvider);
+    _d2ClientJmxManager.setSimpleLoadBalancerState(state);
+
+    SimpleLoadBalancer balancer =
+        new SimpleLoadBalancer(state, _lbTimeout, _lbTimeoutUnit, executorService, _failoutConfigProviderFactory);
```

Comment:
so in DualRead mode, the client will have two simpleLoadBalancerState (one for old zk, one for xds). Both actively create and refresh tracker clients when d2 service properties have updates. Since the xds one will not really be used to send a request under DualRead mode (only for comparing the data), can we have a check and only really create the simple load balancer and its state under NEW_LB_ONLY mode? especially given that we have seen issues with some clients having memory peaks when refreshing tracker clients.

Reply:
as discussed in the standup meeting, we can verify that the xDS LB state shares the existing connections that ZK LB state has, and minimal memory/CPU overhead is added (by enabling dual read on toki and checking its related metrics).